bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/link/LinkBuilderImpl.java

```diff
@@ -0,0 +1,282 @@
+ ~
+ ~     http://www.apache.org/licenses/LICENSE-2.0
+ ~
+ ~ Unless required by applicable law or agreed to in writing, software
+ ~ distributed under the License is distributed on an "AS IS" BASIS,
+ ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ ~ See the License for the specific language governing permissions and
+ ~ limitations under the License.
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
+package com.adobe.cq.wcm.core.components.internal.link;
+
+import java.util.List;
+import java.util.Map;
+import java.util.HashMap;
+import java.util.Optional;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.sling.api.SlingHttpServletRequest;
+import org.apache.sling.api.resource.Resource;
+import org.apache.sling.api.resource.ValueMap;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import com.adobe.cq.wcm.core.components.commons.link.Link;
+import com.adobe.cq.wcm.core.components.commons.link.LinkBuilder;
+import com.adobe.cq.wcm.core.components.services.link.PathProcessor;
+import com.day.cq.dam.api.Asset;
+import com.day.cq.wcm.api.Page;
+import com.day.cq.wcm.api.PageManager;
+
+import static com.adobe.cq.wcm.core.components.commons.link.Link.PN_LINK_ACCESSIBILITY_LABEL;
+import static com.adobe.cq.wcm.core.components.commons.link.Link.PN_LINK_TARGET;
+import static com.adobe.cq.wcm.core.components.commons.link.Link.PN_LINK_TITLE_ATTRIBUTE;
+import static com.adobe.cq.wcm.core.components.internal.Utils.resolveRedirects;
+import static com.adobe.cq.wcm.core.components.internal.link.LinkHandlerImpl.VALID_LINK_TARGETS;
+import static com.adobe.cq.wcm.core.components.internal.link.LinkImpl.ATTR_ARIA_LABEL;
+import static com.adobe.cq.wcm.core.components.internal.link.LinkImpl.ATTR_TARGET;
+import static com.adobe.cq.wcm.core.components.internal.link.LinkImpl.ATTR_TITLE;
+
+public class LinkBuilderImpl implements LinkBuilder {
+
+    public static final String HTML_EXTENSION = ".html";
+
+    SlingHttpServletRequest request;
+    List<PathProcessor> pathProcessors;
+    boolean shadowingDisabled;
+
+    Resource resource;
+    String linkUrl;
+    Map<String, String> linkAttributes = new HashMap<>();
+    Object reference;
+
+    public LinkBuilderImpl(Resource res, SlingHttpServletRequest req, List<PathProcessor> pathProcs, boolean shadowingDisabled) {
+        if (res != null) {
+            resource = res;
+            ValueMap props = resource.getValueMap();
+            linkUrl = props.get(Link.PN_LINK_URL, String.class);
+
+            String linkTarget = props.get(PN_LINK_TARGET, String.class);
+            if (StringUtils.isNotEmpty(linkTarget)) {
+                setLinkTarget(linkTarget);
+            }
+            String linkAccessibilityLabel = props.get(PN_LINK_ACCESSIBILITY_LABEL, String.class);
```

Comment:
PN_LINK_ACCESSIBILITY_LABEL and PN_LINK_TITLE_ATTRIBUTE are marked as deprecated - should they still be used in new code? (unfortunately it's not documented why they are deprecated)

Reply:
understood, ok. than we need this here for backward compatibility.